src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGen.scala

```diff
@@ -732,8 +732,22 @@ class CodeGen (val target: Target) {
         /* run method */
         generateRunMethod(newClass, translationContext, stubType)
 
+        // need to gather the types of the main contract constructor in order to correctly deserialize arguments
+        var constructorTypes: List[ObsidianType] = List.empty
+        for (decl <- translationContext.contract.declarations) {
```

Comment:
This outer loop is kind of weird because you want to assume there is only one constructor, but the code actually corresponds with concatenating arguments from ALL constructors, which doesn't make any sense. How about stopping after you find the first one?